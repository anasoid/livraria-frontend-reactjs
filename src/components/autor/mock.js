export const autores = {
  "autores": [
    {'nome': 'fulano', 'email': 'fulano@fulano.com'},
    {'nome': 'ciclano', 'email': 'ciclano@ciclano.com'}
  ]
};