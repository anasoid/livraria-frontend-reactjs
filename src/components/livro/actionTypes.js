export const GET_LIVROS = 'GET_LIVROS';
export const ADD_LIVRO = 'ADD_LIVRO';
export const REMOVE_LIVRO = 'REMOVE_LIVRO';