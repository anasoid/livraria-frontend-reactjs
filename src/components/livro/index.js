import React from 'react';
import LivroContainer from './livro-container';

const Livro = () => {
  return (
    <LivroContainer/>
  );
}

export default Livro;