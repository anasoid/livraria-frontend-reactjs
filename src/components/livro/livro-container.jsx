import React from 'react';
import LivroTable from './livro-table';

const LivroContainer = () => {
  return (
    <div>
      <h1>Novo Livro</h1>
      <form>
        <label htmlFor={'livroTitulo'}>Titulo: </label>
        <input id={'livroTitulo'} name={'titulo'}/>
        <br/>
        <label htmlFor={'livroIsbn'}>ISBN: </label>
        <input id={'livroIsbn'} name={'isbn'}/>
        <br/>
        <label htmlFor={'livroPreco'}>Preço: </label>
        <input id={'livroPreco'} name={'preco'}/>
        <br/>
        <label htmlFor={'livroDataLancamento'}>Data de Lançamento: </label>
        <input id={'livroDataLancamento'} name={'dataLancamento'}/>
        <br/>
        <label htmlFor={'livroAutores'}>Selecione o Autor: </label>
        <select id={'livroAutores'} name={'selectAutor'}></select>
        <button id={'btnAddAutor'} type={'button'}>Adicionar Autor</button>
        <br/>
        <label>Autores: </label>
        <ul>
          <li></li>
        </ul>
        <br/>
        <button id={'btnGravar'} type={'submit'}>Gravar Livro</button>
      </form>
      <br/>
      <LivroTable/>
    </div>
  );
}

export default LivroContainer;