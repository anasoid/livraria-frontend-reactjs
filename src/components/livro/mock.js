export const livros = {
  livros: [
    {titulo: 'reactjs', isbn: '12345678', preco: 20.0, dtLancamento: '11/05/2018', autores: [{nome: 'fulano', email: 'fulano@email.com'}]},
    {titulo: 'react native', isbn: '87654321', preco: 15.0, dtLancamento: '20/08/2018', autores: [{nome: 'ciclano', email: 'ciclano@email.com'}]}
  ]
}