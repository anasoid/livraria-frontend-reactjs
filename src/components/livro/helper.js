import * as mockResponse from './mock';

export default class Helper {

  getLivrosList() {
    return new Promise(resolve => {
      setTimeout(resolve(mockResponse.livros.livros), 500);
    });
  }

  saveLivro(livro) {
    return new Promise( resolve => {
      mockResponse.livros.livros.push(livro);
      setTimeout(resolve({status: 201}), 500);
    });
  }

  removeLivro(livro) {
    return new Promise(resolve => {
      const index = mockResponse.livros.livros.indexOf(livro);
      mockResponse.livros.livros.splice(index, 1);
      setTimeout(resolve({status: 200}), 500);
    });
  }

}