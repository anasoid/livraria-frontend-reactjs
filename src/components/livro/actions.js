import {GET_LIVROS, ADD_LIVRO, REMOVE_LIVRO} from '../livro/actionTypes';
import Helper from '../livro/helper';

export const getLivrosList = () => {
  const helper = new Helper();
  return dispatch => {
    return helper.getLivrosList()
      .then((data) => {
        return dispatch({type: GET_LIVROS, payload: data});
      }).catch();
  };
};

export const saveLivro = livro => {
  const helper = new Helper();
  return dispatch => {
    return helper.saveLivro(livro)
      .then(() => {
        return dispatch({type: ADD_LIVRO, payload: livro});
      })
  };
};

export const removeLivro = livro => {
  const helper = new Helper();
  return dispatch => {
    return helper.removeLivro(livro)
      .then(() => {
        return dispatch({type: REMOVE_LIVRO, payload: livro});
      });
  };
};