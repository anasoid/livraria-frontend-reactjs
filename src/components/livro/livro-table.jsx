import React, {Component} from 'react';
import {connect} from 'react-redux';
import {bindActionCreators} from 'redux';
import {getLivrosList, removeLivro, selectLivro} from './actions';

export class LivroTable extends Component {

  componentDidMount() {
    this.props.getLivrosList();
  }

  render() {
    return(
      <div>
        <table>
          <caption>Livros</caption>
          <thead>
          <tr>
            <th>Titulo</th>
            <th>ISBN</th>
            <th>Preço</th>
            <th>Data Lançamento</th>
          </tr>
          </thead>
          <tbody>
          {
            this.props.livros.map((livro, index) => {
              return (
                <tr key={index}>
                  <td>{livro.titulo}</td>
                  <td>{livro.isbn}</td>
                  <td>{livro.preco}</td>
                  <td>{livro.dataLancamento}</td>
                  <td><button onClick={() => this.props.selectLivro(index)} >alterar</button></td>
                  <td><button onClick={() => this.props.removeLivro(livro)} >remover</button></td>
                </tr>
              );
            })
          }
          </tbody>
        </table>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  livros: state.livro.livros
});

const mapDispatchToProps = dispatch => bindActionCreators({
  getLivrosList,
  removeLivro,
  selectLivro
}, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(LivroTable);