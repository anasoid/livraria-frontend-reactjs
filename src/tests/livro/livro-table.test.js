import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import {livros} from '../../components/livro/mock';
import {LivroTable} from '../../components/livro/livro-table';

const getLivrosListMock = jest.fn();
const removeLivroMock = jest.fn();
const component = (
  <LivroTable
    livros={livros.livros}
    getLivrosList={getLivrosListMock}
    removeLivro={removeLivroMock}/>
);

describe('test LivroTable components', () => {

  let wrapper;

  beforeEach(() => {
    wrapper = mount(component);
  });

  it('should renders without crashing', () => {
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
    expect(getLivrosListMock).toBeCalled();
  });

  it('should be table with columns Titulo, ISBN, Preço, Data Lançamento', () => {
    const table = wrapper.find('table');
    const thead = table.find('thead').find('tr').find('th');
    expect(table.find('caption')).toHaveText('Livros');
    expect(thead.at(0)).toHaveText('Titulo');
    expect(thead.at(1)).toHaveText('ISBN');
    expect(thead.at(2)).toHaveText('Preço');
    expect(thead.at(3)).toHaveText('Data Lançamento');
  });

  it('should be call props removeLivro when click button remover', () => {
    const rowsTBody = wrapper.find('table').find('tbody').find('tr').find('td');
    const removeButton = rowsTBody.find('button').at(1);

    removeButton.simulate('click');

    expect(removeLivroMock).toBeCalled();
  });

});