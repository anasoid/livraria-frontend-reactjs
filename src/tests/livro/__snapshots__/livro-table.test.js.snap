// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test LivroTable components should renders without crashing 1`] = `
<div>
  <table>
    <caption>
      Livros
    </caption>
    <thead>
      <tr>
        <th>
          Titulo
        </th>
        <th>
          ISBN
        </th>
        <th>
          Preço
        </th>
        <th>
          Data Lançamento
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          reactjs
        </td>
        <td>
          12345678
        </td>
        <td>
          20
        </td>
        <td>
          11/05/2018
        </td>
        <td>
          alterar
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            remover
          </button>
        </td>
      </tr>
      <tr>
        <td>
          react native
        </td>
        <td>
          87654321
        </td>
        <td>
          15
        </td>
        <td>
          20/08/2018
        </td>
        <td>
          alterar
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            remover
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
